File: src/main/java/uk/co/ben_gibson/git/link/Git/File.java
package uk.co.ben_gibson.git.link.Git;

import org.jetbrains.annotations.NotNull;

public class File
{
    private final String path; // A path from the root of the repository.
    private final String name;

    public File(@NotNull String path, @NotNull final String name)
    {
        if (path.startsWith("/")) {
            path = path.substring(1);
        }

        this.path = path;
        this.name = name;
    }

    @NotNull
    public String directoryPath()
    {
        return path.substring(0, (path.length() - name().length()));
    }

    @NotNull
    public String path()
    {
        return path;
    }

    @NotNull
    public String name()
    {
        return name;
    }
}

File: src/main/kotlin/uk/co/ben_gibson/git/link/git/File.kt
package uk.co.ben_gibson.git.link.git

import com.intellij.openapi.vfs.VirtualFile
import git4idea.repo.GitRepository

data class File(
    val name: String,
    val isDirectory: Boolean,
    val path: String,
    val isRoot: Boolean
) {
    companion object {
        fun forRepository(file: VirtualFile, repository: GitRepository): File {
            val isRoot = file.path == repository.root.path
            val relativePath = if (isRoot) {
                ""
            } else {
                file.path.substring(repository.root.path.length)
                    .removeSuffix("/${file.name}")
                    .trim('/')
            }

            return File(
                name = file.name,
                isDirectory = file.isDirectory,
                path = relativePath,
                isRoot = isRoot
            )
        }
    }
}
